compiler/generator_dart.c

```diff
@@ -0,0 +1,1274 @@
+        write_str(g, g->failure_str);
+        write_newline(g);
+    }
+    write_margin(g);
+    switch (g->failure_label) {
+        case x_return:
+            write_string(g, "return false;");
+            g->unreachable = true;
+            break;
+        default:
+            write_string(g, "throw BreakLab");
+            write_int(g, g->failure_label);
+            write_string(g, "();");
+            g->unreachable = true;
+    }
+    write_newline(g);
+}
+
+static void write_failure_if(struct generator * g, char * s, struct node * p) {
+
+    writef(g, "~Mif (", p);
+    writef(g, s, p);
+    writef(g, ")~N", p);
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+}
+
+/* if at limit fail */
+static void write_check_limit(struct generator * g, struct node * p) {
+
+    if (p->mode == m_forward) {
+        write_failure_if(g, "cursor >= limit", p);
+    } else {
+        write_failure_if(g, "cursor <= limit_backward", p);
+    }
+}
+
+/* Formatted write. */
+static void writef(struct generator * g, const char * input, struct node * p) {
+    int i = 0;
+    int l = strlen(input);
+
+    while (i < l) {
+        int ch = input[i++];
+        if (ch != '~') {
+            write_char(g, ch);
+            continue;
+        }
+        switch (input[i++]) {
+            default: write_char(g, input[i - 1]); continue;
+            case 'C': write_comment(g, p); continue;
+            case 'f': write_block_start(g);
+                      write_failure(g);
+                      g->unreachable = false;
+                      write_block_end(g);
+                      continue;
+            case 'M': write_margin(g); continue;
+            case 'N': write_newline(g); continue;
+            case '{': write_block_start(g); continue;
+            case '}': write_block_end(g); continue;
+            case 'S': write_string(g, g->S[input[i++] - '0']); continue;
+            case 'B': write_b(g, g->B[input[i++] - '0']); continue;
+            case 'I': write_int(g, g->I[input[i++] - '0']); continue;
+            case 'V': write_varref(g, g->V[input[i++] - '0']); continue;
+            case 'W': write_varname(g, g->V[input[i++] - '0']); continue;
+            case 'L': write_literal_string(g, g->L[input[i++] - '0']); continue;
+            case '+': g->margin++; continue;
+            case '-': g->margin--; continue;
+            case 'n': write_string(g, g->options->name); continue;
+        }
+    }
+}
+
+static void w(struct generator * g, const char * s) {
+    writef(g, s, 0);
+}
+
+static void generate_AE(struct generator * g, struct node * p) {
+    const char * s;
+    switch (p->type) {
+        case c_name:
+            write_varref(g, p->name); break;
+        case c_number:
+            write_int(g, p->number); break;
+        case c_maxint:
+            write_string(g, "9007199254740991"); break;
+        case c_minint:
+            write_string(g, "-9007199254740991"); break;
```

Comment:
I assume this means Dart uses C `double` for its numbers.

We should probably try to clarify the docs about what `maxint` and `minint` are exactly.  Currently we say "the maximum positive number" and "the minimum negative number", but you can have a larger positive number than `maxint` in a Snowball integer for Dart.  Setting `maxint` to `HUGEVAL` would fulfil the letter of the definition, but causes problems for code that (quite resaonably) expects to be able to represent any integer between `minint` and `maxint`,

I believe Javascript uses `double` for its numbers, and there we use `(-1>>>1)` for `maxint` and `(~(-1>>>1))` for `minint` which give the limits for a signed 32 bit integer (at least with current node on x86-64) but you can go above `maxint`:

```
$ node -e 'console.log(-1>>>1);console.log(~(-1>>>1));console.log((-1>>>1)+1)'
2147483647
-2147483648
2147483648
```

For Javascript we could probably enforce these as actual limits by wrapping every integer expression as `(expr)|0`:

```
$ node -e 'console.log(((-1>>>1)+1)|0)'
-2147483648
```

I tend to think that not actually a worthwhile thing to do though.  None of the current stemmers we ship actually use `minint` or `maxint`, but the way in which they might be useful in Snowball's intended domain of use would be to serve as a `not found` type value, and in that situation a variable being able to exceed the limit isn't actually a major problem - it just need to be clearly documented so that people test with `<` and `>=` instead of `!=` and `==`.

Reply:
I interpret `minint` and `maxint` as the limits beyond which integer arithmetic is not guaranteed to work as mathematically expected. For example, Dart’s `maxint` should be `2 ** 53` because beyond that not every integer can be represented.

Under that interpretation, the current JavaScript implementation’s `minint` and `maxint` are too small: they should be `Number.MIN_SAFE_INTEGER - 1` and `Number.MAX_SAFE_INTEGER + 1`